### PR TITLE
expose session and transaction helpers to other services

### PR DESCRIPTION
**Pull request by bergquist**
to enable other services to use the database

### CODE REVIEW

File: pkg/services/sqlstore/sqlstore.go [hunk 1]
```diff
@@ -56,6 +56,64 @@ type SqlStore struct {
...
+	sess, err := startSession(ctx, ss.engine, false)
+	if err != nil {
+		return err
+	}
+
+	return callback(sess)
+}
+
+// WithTransactionalDbSession calls the callback with an session within a transaction
+func (ss *SqlStore) WithTransactionalDbSession(ctx context.Context, callback dbTransactionFunc) error {
```
Comment (torkelo) [inline pkg/services/sqlstore/sqlstore.go]: this name seems a bit much, can't this use InTransaction?

File: pkg/services/sqlstore/sqlstore.go [hunk 2]
```diff
@@ -56,6 +56,64 @@ type SqlStore struct {
 	skipEnsureAdmin bool
 }
 
+// NewSession returns a new DBSession
+func (ss *SqlStore) NewSession() *DBSession {
```
Comment (torkelo) [inline pkg/services/sqlstore/sqlstore.go]: can't this be in the session.go file?

File: pkg/services/sqlstore/sqlstore.go [hunk 3]
```diff
@@ -56,6 +56,64 @@ type SqlStore struct {
...
+
+	return callback(sess)
+}
+
+// WithTransactionalDbSession calls the callback with an session within a transaction
+func (ss *SqlStore) WithTransactionalDbSession(ctx context.Context, callback dbTransactionFunc) error {
+	return ss.inTransactionWithRetryCtx(ctx, callback, 0)
+}
+
+func (ss *SqlStore) inTransactionWithRetryCtx(ctx context.Context, callback dbTransactionFunc, retry int) error {
```
Comment (torkelo) [inline pkg/services/sqlstore/sqlstore.go]: Does this not already exist in sqlstore/transactions, https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/transactions.go#L16

**Comment by bergquist** [comment]
Merging this. This is a messy chance but we need it so other services can perform database tasks

We will continue the discussion for how we can improve this in https://github.com/grafana/grafana/issues/12354

**Pull request merged**